### ISSUE TITLE
Spring-Retry @Retryable not working with Spring-data Repositories

### ISSUE THREAD

**Issue by jopuneet** [spring-projects/spring-retry#214]
I was trying @Retryable with Spring-data and testing it out using Mockito @MockBean but the retry for the repository is not working and it was always called only once.

```
@Repository
public interface TestRepository extends CrudRepository<Test, String> {

  @Retryable(value = { Exception.class }, maxAttempts = 4, backoff = @Backoff(delay = 2000))
  List<Test> findAll();
}
```
```
when(testRepository.findAll())
            .thenThrow(new RecoverableDataAccessException("test RecoverableDataAccessException 1"))
            .thenThrow(new RecoverableDataAccessException("test RecoverableDataAccessException 2"))
            .thenReturn(Lists.newArrayList(test));
verify(testRepository, times(3)).findAll(any());
```
```
org.mockito.exceptions.verification.TooFewActualInvocations: 
testRepository bean.findAll(<any string>);
Wanted 3 times:
-> at com.*.findAllTest(TestIT.java:10)
But was 1 time:
```

Referring to this [Stackoverflow answer](https://stackoverflow.com/questions/42730104/retryable-doesnt-work-in-spring-data-repository-interface)

Spring-boot: 2.3.0.RELEASE
Spring-Retry: 1.2.5.RELEASE
Spring-data: 2.3.0.RELEASE

**Comment by garyrussell** [comment]
If you are mocking the `TestRepository` then it's a mock instead of a Spring Bean so, of course, retry won't work; Mockito doesn't know anything about Spring.

**Comment by Puneetjoshi24** [comment]
@garyrussell I am using `@MockBean` to mock the `TestRepository` not `@Mock` and `@MockBean` provide spring context with the mock. As referenced here [Baeldung MockBean](https://www.baeldung.com/java-spring-mockito-mock-mockbean#:~:text=We%20can%20use%20the%20%40MockBean,type%20in%20the%20application%20context.&text=When%20we%20use%20the%20annotation,be%20injected%20into%20the%20field.) 

If spring retry will not work with Mock here, what is the way to test Spring-retry over Spring-Jpa? We already have a question on [Stackoverflow](https://stackoverflow.com/questions/42730104/retryable-doesnt-work-in-spring-data-repository-interface) .

**Comment by Puneetjoshi24** [comment]
Also, I tested this in real-time (not only in the test case), Spring-retry is not working on Spring-data repository. On retrying on any exception and returning a value from `@Query()` which is not expected, it is not retrying rather it is throwing this: 

`org.springframework.beans.factory.BeanNotOfRequiredTypeException:  Bean named 'testRepository' is expected to be of type 'com.*.TestRepository' but was actually of type 'com.sun.proxy.$Proxy223'`

**Comment by garyrussell** [comment]
For the "real" case, it is because spring-retry is wrapping the JPA proxy in another proxy that only implements `Retryable`.

So, yes, `@Retryable` is not supported directly on JPA repositories.

**Comment by garyrussell** [comment]
Here is a work around - remove `@Retryable` and wire an interceptor directly into the JPA proxy.

```java
@SpringBootApplication
public class Srgh214Application {


	private static final Logger log = LoggerFactory.getLogger(Srgh214Application.class);


	public static void main(String[] args) {
		SpringApplication.run(Srgh214Application.class, args);
	}

	@Bean
	public ApplicationRunner runner(TestRepository repo) {
		return args -> {
			repo.findAll();
		};
	}

	@Bean
	static BPP bpp() {
		return new BPP();
	}

}

class BPP implements BeanPostProcessor, Ordered {

	private static final Logger log = LoggerFactory.getLogger(BPP.class);

	@Override
	public int getOrder() {
		return Integer.MAX_VALUE;
	}

	@Override
	@Nullable
	public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
		if (bean instanceof TestRepository) {
			Advised advised = (Advised) bean;
			RetryOperationsInterceptor interceptor = RetryInterceptorBuilder.stateless()
					.maxAttempts(4)
					.backOffOptions(2000, 1.0, 2000)
					.recoverer((args, cause) -> {
						log.info("retries exhausted: " + cause.getMessage());
						throw (RuntimeException) cause;
					})
					.build();
			advised.addAdvice(0, interceptor);
		}
		return bean;
	}

}
```

**Comment by Puneetjoshi24** [comment]
@garyrussell Thanks this works perfectly fine. 
We can close this issue if this is expected and we will not add any capability on Spring-retry to get integrated with Spring-Data Repositories

**Comment by kaganuk** [comment]
adding support for that can be useful :)

**Comment by kunaldoshi98** [comment]
@garyrussell, the code you mentioned, will it be helpful to add retries on JPA repo which is doing DB calls present in the main

**Comment by mp911de** [comment]
What would be the best place for a generic Spring Data integration? A `BeanPostProcessor` or similar can add another interceptor to Spring Data's `ProxyFactory`:

```java
RepositoryFactoryBeanSupport factoryBean = …;
factoryBean.addRepositoryFactoryCustomizer(repositoryFactory -> {
	
	repositoryFactory.addRepositoryProxyPostProcessor((proxyFactory, repositoryInformation) -> {
		proxyFactory.addAdvisor(…);
	});
});
```

Adding Spring Retry logic including annotation scanning to Spring Data isn't probably the best approach as we would duplicate already existing code to some extent.

**Comment by garyrussell** [comment]
spring-retry currently uses an `AbstractAutoProxyCreator` which is a BPP (from spring-aop) that unconditionally wraps the "bean" in another proxy that only implements `Retryable`.

Interestingly

```java
	@Bean
	public ApplicationRunner runner(TestRepository repo) {
		return args -> {
			repo.findAll();
		};
	}
```

Gets a repo that is injected before the retry BPP has been applied (not sure why yet), and 

```java
	@Bean
	public ApplicationRunner runner2(ApplicationContext context) {
		return args -> {
			context.getBean(TestRepository.class).findAll();
		};
	}
```

fails because by now the proxy only implements `Retryable`.

We can somehow fix it by adding the retry interceptor advice to the existing proxy, but we would also need to prevent the auto proxy BPP from wrapping it in another proxy.

I don't think this will need any changes in spring-jpa.

I'll dig around some more...

**Comment by garyrussell** [comment]
I have found a solution; will submit a PR soon.

However, I have discovered that the root cause of the problem is Boot sets `proxyTargetClass` to `true` by default.

It works for me with

     spring.aop.proxy-target-class=false

When it is true, the auto-proxy feature fails to copy the inner proxy's interface to the outer proxy; when it is false, the interfaces are copied.

My solution (for when it is true) is to re-create a single proxy with a merged set of interfaces and advisors.

**Comment by artembilan** [comment]
> a single proxy with a merged set of interfaces and advisors.

And that is actually what we always do in Spring Integration. If that is already a proxy we add our advises to it otherwise create:
```
if (AopUtils.isAopProxy(source)) {
	Advised advised = (Advised) source;
	chain.forEach(advice -> advised.addAdvisor(advice));
}
else {
	ProxyFactory proxyFactory = new ProxyFactory(source);
	chain.forEach(advice -> proxyFactory.addAdvisor(advice));
	source = proxyFactory.getProxy(getBeanClassLoader());
}
```

**Comment by garyrussell** [comment]
Yes; but it's more complicated here, because Spring AOP auto-proxy creates a new proxy unconditionally. It only copies the interfaces to the outer proxy if `proxyTargetClass` is false (Boot sets it to true by default).

**Comment by elafontaine** [comment]
Hi, how are we to use the Retryable with Repository today ?  is it still the same as https://github.com/spring-projects/spring-retry/issues/214#issuecomment-685942868

Just got bitten by it.

**Comment by garyrussell** [comment]
Fixed in Spring Framework 5.3.9 https://github.com/spring-projects/spring-framework/issues/27044
